### PR TITLE
feat(211): Evolution on Services types verification for ControlBlock …

### PR DESCRIPTION
**Pull request by SaintierFr**
Evolution on Service types verification for ControlBlock and DataSet creation

Enable preview for java 17 features
Add enable-preview in properties for Java 17 switch (needed for tests)
Upgrade Cache Action to v3

### CODE REVIEW

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java [hunk 1]
```diff
@@ -172,91 +175,195 @@ void testMatches() throws Exception {
...
+        // When Then
+        assertDoesNotThrow(() -> iAdapter.createDataSet(dataSetInfo));
         assertIsMarshallable(scd);
     }
+    @Test
+    void createDataSet_should_throw_ScdException_when_DataSetInfo_is_incomplete() throws Exception {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        SclRootAdapter sclRootAdapter = new SclRootAdapter(scd);
+        IEDAdapter iAdapter = assertDoesNotThrow(() -> sclRootAdapter.getIEDAdapterByName("IED_NAME"));
```
Comment (gleizesDor) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:204]: many asserts in this test, you should remove this one as if it throw an exception, the test will fail

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java [hunk 2]
```diff
@@ -172,91 +175,195 @@ void testMatches() throws Exception {
...
+
+        TDataSet tDataSet = new TDataSet();
+        tDataSet.setName("dataset");
+        TFCDA tfcda = new TFCDA();
+        tfcda.setFc(TFCEnum.ST);
+        tDataSet.getFCDA().add(tfcda);
+        DataSetInfo dataSetInfo = DataSetInfo.from(tDataSet);
+        dataSetInfo.setHolderIEDName("IED_NAME");
+
+        assertThrows(ScdException.class, () -> iAdapter.createDataSet(dataSetInfo));
```
Comment (gleizesDor) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:214]: assertJ ?

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java [hunk 3]
```diff
@@ -172,91 +175,195 @@ void testMatches() throws Exception {
...
-    void testHasDataSetCreationCapability() throws Exception {
+    void createDataSet_should_throw_ScdException_when_DataSetInfo_is_empty() throws Exception {
+        // Given
         SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
         SclRootAdapter sclRootAdapter = new SclRootAdapter(scd);
         IEDAdapter iAdapter = assertDoesNotThrow(() -> sclRootAdapter.getIEDAdapterByName("IED_NAME"));
 
-        assertFalse(iAdapter.hasDataSetCreationCapability());
+        // When Then
+        assertThrows(ScdException.class, () -> iAdapter.createDataSet(new DataSetInfo()));
```
Comment (gleizesDor) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:225]: assertJ ?

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java [hunk 4]
```diff
@@ -172,91 +175,195 @@ void testMatches() throws Exception {
...
+        assertThrows(ScdException.class, () -> iAdapter.createDataSet(dataSetInfo));
+    }
 
     @Test
-    void testHasDataSetCreationCapability() throws Exception {
+    void createDataSet_should_throw_ScdException_when_DataSetInfo_is_empty() throws Exception {
+        // Given
         SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
         SclRootAdapter sclRootAdapter = new SclRootAdapter(scd);
         IEDAdapter iAdapter = assertDoesNotThrow(() -> sclRootAdapter.getIEDAdapterByName("IED_NAME"));
```
Comment (gleizesDor) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:222]: assertDoesNotThrow not necessary

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java [hunk 5]
```diff
@@ -172,91 +175,195 @@ void testMatches() throws Exception {
...
 
-        TServices tServices = new TServices();
-        iAdapter.getCurrentElem().setServices(tServices);
-        assertFalse(iAdapter.hasDataSetCreationCapability());
+    @Test
+    void hasDataSetCreationCapability_should_return_false_when_no_existing_services_attribute() throws Exception {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        SclRootAdapter sclRootAdapter = new SclRootAdapter(scd);
+        IEDAdapter iAdapter = assertDoesNotThrow(() -> sclRootAdapter.getIEDAdapterByName("IED_NAME"));
```
Comment (gleizesDor) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:233]: assertDoesNotThrow not necessary

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java [hunk 6]
```diff
@@ -172,91 +175,195 @@ void testMatches() throws Exception {
...
 
-        TGSESettings tgseSettings = new TGSESettings();
-        tServices.setGSESettings(tgseSettings);
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("provideFalseTServicesCases")
+    void hasDataSetCreationCapability_should_return_false_when_wrong_attribute(String testCase, TServices tServices, TServiceSettings serviceSettings) throws Exception {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        SclRootAdapter sclRootAdapter = new SclRootAdapter(scd);
+        IEDAdapter iAdapter = assertDoesNotThrow(() -> sclRootAdapter.getIEDAdapterByName("IED_NAME"));
```
Comment (gleizesDor) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:245]: assertDoesNotThrow not necessary

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java [hunk 7]
```diff
@@ -172,91 +175,195 @@ void testMatches() throws Exception {
...
+        );
+    }
 
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("provideTrueTServicesCases")
+    void hasDataSetCreationCapability_should_return_true_when_attribute_exists(String testCase, TServices tServices, TServiceSettings serviceSettings) throws Exception {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        SclRootAdapter sclRootAdapter = new SclRootAdapter(scd);
+        IEDAdapter iAdapter = assertDoesNotThrow(() -> sclRootAdapter.getIEDAdapterByName("IED_NAME"));
```
Comment (gleizesDor) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:301]: assertDoesNotThrow not necessary

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapter.java [hunk 8]
```diff
@@ -359,38 +403,37 @@ public ControlBlock<? extends ControlBlock> createControlBlock(ControlBlock<? ex
         TServices tServices = currentElem.getServices();
         TServiceSettingsNoDynEnum cbNameAtt = controlBlock.getControlBlockServiceSetting(tServices);
```
Comment (massifben) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapter.java:404]: It seems that the wrong XPath is checked.
The issue requires to check `/SCL/IED/AccessPoint@name="XXXXXX"/Services/YYYSettings/@cbName` bu what is checked here is `/SCL/IED/Services/YYYSettings/@cbName`

**Comment by massifben** [comment]
I don't think it is a good idea to use preview features of Java.
We should stick to stable features of Java 17.

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapter.java [hunk 9]
```diff
@@ -317,38 +334,65 @@ public void createDataSet(DataSetInfo dataSetInfo) throws ScdException {
...
+            case TLogSettings ignored -> isSettingsDatSetConfOrDyn(getLogSettingsEnumStream());
+            case TReportSettings ignored -> isSettingsDatSetConfOrDyn(getReportSettingsEnumStream());
+            case TGSESettings ignored -> isSettingsDatSetConfOrDyn(getGseSettingsEnumStream());
+            case TSMVSettings ignored -> isSettingsDatSetConfOrDyn(getSmvSettingsEnumStream());
+            default -> false;
+        };
+    }
+
+    private Stream<TServices> getServicesStream() {
+        return currentElem.getAccessPoint().stream()
```
[the remark this answers is not in the document]
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapter.java:360]: Fixed : move `hasDataSetCreationCapability` and `hasControlBlockCreationCapability` to `LDeviceAdapter`.

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapter.java (hunk 8 above)
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapter.java:404]: Fixed.

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java (hunk 1 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:204]: OK

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java (hunk 2 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:214]: Fixed

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java (hunk 4 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:222]: Fixed

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java (hunk 3 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:225]: fixed

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java (hunk 5 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:233]: removed

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java (hunk 6 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:245]: removed

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java (hunk 7 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/IEDAdapterTest.java:301]: removed

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LDeviceAdapterTest.java [hunk 10]
```diff
@@ -173,7 +176,246 @@ void getLNAdaptersInclundigLN0() {
...
+        IEDAdapter iedAdapter = sclRootAdapter.getIEDAdapterByName("IED_NAME");
+        LDeviceAdapter lDeviceAdapter = iedAdapter.getLDeviceAdapterByLdInst("LD_INS1");
+        iedAdapter.getCurrentElem().getAccessPoint().get(0).setServices(new TServices());
+
+        // When
+        boolean hasCapability = lDeviceAdapter.hasControlBlockCreationCapability(serviceSettingsType);
+        //Then
+        assertThat(hasCapability).isFalse();
+    }
+
```
Comment (AliouDIAITE) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LDeviceAdapterTest.java:419]: The case in which ServiceSettingsType = null is it tested?

Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LDeviceAdapterTest.java:433]: No it is missing.
OK, I add a test for hasDataSetCreationCapability and hasControlBlockCreationCapability when ServiceSettingsType == null.
